### ISSUE TITLE
Incorrect `gather_for_metrics` evaluation with multiple GPUs and uneven batch sizes

### ISSUE THREAD

**Issue by activatedgeek** [huggingface/accelerate#1674]
### System Info

```Shell
accelerate version: 0.20.3
OS: Ubuntu 20.04
python version: 3.10.11
numpy version: 1.24.3
torch version: 1.13.1
accelerate's configuration: (see run command)
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Here is the code that fails. This code simply computes a ratio between the `limit / n` which are the arguments of the main function and used via the CLI as.

```python
from tqdm.auto import tqdm
from torch.utils.data import Dataset, DataLoader
from accelerate import Accelerator
import torch


def get_num_workers(num_workers=4):
    num_gpus_per_host = torch.cuda.device_count()
    return (num_workers + num_gpus_per_host - 1) // num_gpus_per_host


def get_loader(dataset, batch_size=128, num_workers=4, accelerator=None, **kwargs):
    num_workers = get_num_workers(num_workers=num_workers)
    loader = DataLoader(
        dataset, batch_size=batch_size, num_workers=num_workers, **kwargs
    )
    if accelerator is not None:
        loader = accelerator.prepare(loader)

    return loader


class TestData(Dataset):
    def __init__(self, n=100):
        self.x = range(1, n + 1)

    def __len__(self):
        return len(self.x)
    
    def __getitem__(self, idx):
        return self.x[idx], self.x[idx]
    

def main(n=100, batch_size=10, limit=50):
    accelerator = Accelerator()

    data = TestData(n=n)
    loader = get_loader(data, batch_size=batch_size, accelerator=accelerator)

    N = torch.tensor(0).long().to(accelerator.device)
    N_acc = torch.tensor(0).long().to(accelerator.device)

    for x, y in tqdm(loader):
        pred = torch.where(x > limit, y + 1, y)

        __N = torch.tensor(x.size(0)).long().to(accelerator.device)
        __N_acc = (pred == y).sum().long()
        (__N, __N_acc) = accelerator.gather_for_metrics((__N, __N_acc))

        N += __N.sum()
        N_acc += __N_acc.sum()

    accelerator.wait_for_everyone()

    if accelerator.is_main_process:
        print(f"Ratio: {N_acc / N:.6f}")
        assert torch.allclose((N_acc / N), torch.tensor(limit / n).to(accelerator.device))


if __name__ == "__main__":
    import fire
    fire.Fire(main)
```

Running the following commands with just a single GPU or 4 GPUs returns the correct answer `0.057200`:
```shell
CUDA_VISIBLE_DEVICES=0 accelerate launch --multi_gpu test.py --n=10000 --limit=572 --batch-size=10
```

```shell
CUDA_VISIBLE_DEVICES=0,1,2,3 accelerate launch --multi_gpu test.py --n=10000 --limit=572 --batch-size=10
```

Batch size 1 also works.

However, changing the batch size to be uneven gives an incorrect answer `0.063196` and the assertion above fails:
```shell
CUDA_VISIBLE_DEVICES=0,1,2,3 accelerate launch --multi_gpu test.py --n=10000 --limit=572 --batch-size=17
```

Setting `accelerator = Accelerator(even_batches=False)` hangs the process towards the end of the dataloader loop in the case of uneven batches.

### Expected behavior

The distributed data loader should work with uneven batch sizes and any number of GPUs.

**Comment by activatedgeek** [comment]
It looks like #982 fixed a slightly different issue.

**Comment by activatedgeek** [comment]
Setting `accelerator = Accelerator(even_batches=False)` hangs the process.

**Comment by activatedgeek** [comment]
Hi @muellerzr, bumping this up in case there is an easy way to get around this issue to discard additional padding for evaluation purposes (I think this behavior doesn't hurt stochastic training). Thank you!

**Comment by sgugger** [comment]
@muellerzr friendly ping.

**Comment by muellerzr** [comment]
Looking into it this week, thanks @activatedgeek

**Comment by activatedgeek** [comment]
Thank you!

**Comment by muellerzr** [comment]
The issue here stems from `__N` having too many items.  As for why, `gather_for_metrics` checks if we have a remainder based on the current number of items in a batch. So, for example say we have `[0,1,2,3,4]` on 2 gpus with a bs of 2 on each, we have a remainder of the number `4`, which would get dropped. In this situation, as you are doing `x.size(0).long()` we don't actually get a tensor of the right length to where we need to drop the remainder.

So here you're not using the API properly essentially. The reason it doesn't bug out before is simply because it's never actually dropped with those extra samples. As it never had enough samples *to be dropped*. Does this make sense? Just check the size of `__N` vs the sizes of `pred` and `y` throughout it all, and you will understand what I mean by this.

**Comment by activatedgeek** [comment]
Ah I see. Thank you for clarifying this. I understand that any object gathered has to be of the same length as the objects from the DataLoader so that it gathers with the correct padding.

I was able to fix it with the following changes instead:
```diff
-    N = torch.tensor(0).long().to(accelerator.device)
-    N_acc = torch.tensor(0).long().to(accelerator.device)
+    acc = []

    for x, y in tqdm(loader):
        pred = torch.where(x > limit, y + 1, y)

+       __acc = (pred == y)
+       (__acc,) = accelerator.gather_for_metrics((__acc,))
+       acc.append(__acc)

-        __N = torch.tensor(x.size(0)).long().to(accelerator.device)
-        __N_acc = (pred == y).sum().long()
-        (__N, __N_acc) = accelerator.gather_for_metrics((__N, __N_acc))

-        N += __N.sum()
-        N_acc += __N_acc.sum()

+    acc = torch.cat(acc, axis=0).float().mean()
```

Is there any other approach that you recommend in case I do not want to pass around large tensors?
